pandas/tslib.pyx

```diff
@@ -2095,7 +2095,7 @@ cpdef array_with_unit_to_datetime(ndarray values, unit, errors='coerce'):
         # if we have nulls that are not type-compat
         # then need to iterate
         try:
-            iresult = values.astype('i8')
+            iresult = values.astype('i8', casting='no')
```

Comment:
I think this should actually be `casting='same_kind'`.  You can also add `copy=False` to avoid a copy in the case where the values are already int64.

```python
In [35]: a = np.array([1, 2, 3], dtype='int32')

In [36]: a.astype('int64', casting='no')
TypeError                                 Traceback (most recent call last)
<ipython-input-36-efb09bba64c8> in <module>()
----> 1 a.astype('int64', casting='no')

TypeError: Cannot cast array from dtype('int32') to dtype('int64') according to the rule 'no'

In [37]: a.astype('int64', casting='same_kind')
Out[37]: array([1, 2, 3], dtype=int64)
```